source/language/minimal_profile.md

```diff
@@ -0,0 +1,130 @@
+# OpenQASM 3.0 `minimal` profile
+
+
+## Overview
+
+This *profile* defines a minimum subset of the OpenQASM 3.0 language
+that implementors are assumed to support. This should be sufficient to
+form a target for experiment authors or circuit libraries who intend
+to express *executable* OpenQASM -- that is, OpenQASM programs that
+may be assumed to be runnable by conforming implementations.
+
+The [OpenQASM specification](https://openqasm.com/) serves as the
+definition of the language. This profile introduces assumptions or
+constraints on top of the language specification, effectively defining
+a subset of valid OpenQASM programs. This profile also gives
+suggestions or guidelines for implementors.
+
+In this document, we introduce a small bit of jargon loosely inspired
+by [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119):
+- `MAY` indicates some aspect of the language whose support is
+  required, and whose usage is consistent with this minimal profile
+- `MAY NOT` indicates some aspect of the language that might be
+  supported by particular implementors, but which implies that the
+  OpenQASM program is not within the purview of this minimal profile
+- `SHOULD` and `SHOULD NOT` indicate some best-effort requirement on
+  behalf of implementors. Typically this is not a constraint on the
+  set of programs that meet the profile, but rather a suggestion as to
+  their meaning or execution.
+
+## This Profile
+
+At a high level, this profile assumes that implementors can execute
+"straight-line circuits" with quantum measurements. We do not consider
+nontrivial control flow or classical data processing. The goal is to
+make this truly minimal: we are ready to assume a certain amount of
+upstream language processing.
+
+##  Language Requirements
+
+1.   MAY NOT use `include` statements, other than `include "stdgates.inc"`
+	-  Why: I am not convinced that `include` is used much. I would
+       also assume that the search path depends on implementation
+       specific details.
+2. MAY NOT declare or use *variables* of any classical type
+	- Why: Implementations are probably quite uneven both in the
+      types, operations, and circumstances in which classical data
+      computed at run-time may be used. It's not needed for many
+      "textbook" quantum circuits.
+3. MAY use `float` literals
+	- Why: needed for parametric gates
+4. MAY NOT use integer, boolean, bit string, or duration literals
+	- Why: These are only usable with other language features I am
+      excluding; hence we can exclude these.
+5. MAY NOT use any other compile-time constant expressions
+	- Why: Reduce barrier to entry for implementors. It's easy to
+      imagine a library (or even a visitor in the OpenQASM repo)
+      folding constants, so we'd end up at the literals.
+6. MAY NOT use qubit registers or classical registers
+	- Why: Register manipulations require bounds checking -- why bother.
```

Comment:
Interesting. I think when one looks at existing tool support for OpenQASM in the wild one sees kind of the opposite, where these tools support qubit and classical registers, but they don't support physical qubits.

Reply:
My starting point here was the spec rather than the ecosystem. From that perspective, qubit registers are more work to implement than physical qubits. Although I mention the bounds checking, there is another small bump in complexity since virtual qubits and qubit registers require managing user defined names.

Reply:
To summarize the recent discussion
- It's not much of a lift to support globally declared qubit registers
- It might not make sense for a minimalistic profile to include both virtual and physical qubits, due to the heterogeneity / language surface area. The rough consensus is that we'd be better off supporting virtual qubits (including qubit registers) and leave physical qubits for a profile that deals with defcals.